shared/data/web_accessible_resources/youtube-iframe-api.js

```diff
@@ -0,0 +1,288 @@
+            // necessary for the onReady event to fire for the video.
+            url.searchParams.set('enablejsapi', '1')
+
+            const videoIframe = document.createElement('iframe')
+            videoIframe.height = parseInt(height, 10) || defaultHeight
+            videoIframe.width = parseInt(width, 10) || defaultWidth
+            videoIframe.src = url.href
+
+            if (target.id) {
+                videoIframe.id = target.id
+            }
+
+            target.replaceWith(videoIframe)
+            target = videoIframe
+
+            target.dispatchEvent(new CustomEvent('ddg-ctp-replace-element'))
+        }
+
+        if (events) {
+            if (events.onReady) {
+                onReadyListenerByVideoElement.set(target, events.onReady)
+            }
+            if (events.onStateChange) {
+                onStateChangeListenerByVideoElement.set(target, events.onStateChange)
+            }
+        }
+
+        mockPlayerByVideoElement.set(target, this)
+        this.playerInfo = {}
+        return this
+    }
+
+    // Stub out the YouTube Iframe API.
+    window.YTConfig = {
+        host: 'https://www.youtube.com'
+    }
+    window.YT = {
+        loading: 1,
+        loaded: 1,
+        Player,
+        PlayerState: {
+            UNSTARTED: -1,
+            ENDED: 0,
+            PLAYING: 1,
+            PAUSED: 2,
+            BUFFERING: 3,
+            CUED: 5
+        },
+        setConfig (config) {
+            for (const key of Object.keys(config)) {
+                window.YTConfig[key] = config[key]
+            }
+        },
+        get () { },
+        ready () { },
+        scan () { },
+        subscribe () { },
+        unsubscribe () { }
+    }
+
+    /**
+     * Load the YouTube Iframe API, replacing the stub.
+     * @return {Promise}
+     *   Promise which resolves after the API has finished loading.
+     */
+    function ensureYouTubeIframeAPILoaded () {
+        if (youTubeIframeAPILoaded) {
+            return Promise.resolve()
+        }
+        if (youTubeIframeAPILoadingPromise) {
+            return youTubeIframeAPILoadingPromise
+        }
+
+        const loadingPromise = new Promise((resolve, reject) => {
+            // The YouTube Iframe API calls the `onYouTubeIframeAPIReady`
+            // function to signal that it is ready for use by the website.
+            window.onYouTubeIframeAPIReady = resolve
+        }).then(() => {
+            window.onYouTubeIframeAPIReady = realOnYouTubeIframeAPIReady
+            RealYTPlayer = window.YT.Player
+            youTubeIframeAPILoaded = true
+            youTubeIframeAPILoadingPromise = null
+        })
+
+        // Delete the stub `YT` Object, since its presence will prevent the
+        // YouTube Iframe API from loading.
+        // Note: There's a chance that website will attempt to reference the
+        //       `YT` Object inbetween now and when the script has loaded. This
+        //       is unfortunate but unavoidable.
```

Comment:
If this becomes a problem we might be able to use defineProperty to detect who the caller is.

Reply:
Good idea, I'll bear that in mind.